feat: allow four digit versioning by way of revision field

mikefero opened the pull request:
This feature adds a four digit version parsing while still allowing strict semver. Kong Gateway and Kong Gateway OSS use different versioning systems; Kong Gateway doesn't follow strict semver. As Kong Gateway and Kong Gateway OSS are released the major.minor.patch versions are in lock step; however, Kong Gateway may have additional revisions that are based of the OSS upstream.

Comparison and version ranges can be mixed between 3 and 4 digit versions. When comparing 3 digit version to a 4 digit version only the major.minor.patch-prerelease+buildmetadata will apply. This allows the comparison of Kong Gateway and Kong Gateway OSS versions.

mikefero committed:
feat: allow four digit versioning by way of revision field

This feature adds a four digit version parsing while still allowing
strict semver. Kong Gateway and Kong Gateway OSS use different
versioning systems; Kong Gateway doesn't follow strict semver. As Kong
Gateway and Kong Gateway OSS are released the major.minor.patch
versions are in lock step; however, Kong Gateway may have additional
revisions that are based of the OSS upstream.

Comparison and version ranges can be mixed between 3 and 4 digit
versions. When comparing 3 digit version to a 4 digit version only the
major.minor.patch-prerelease+buildmetadata will apply. This allows the
comparison of Kong Gateway and Kong Gateway OSS versions.

The changes in the root and the v4 directory are identical and the
go.mod file in the root has been updated to match the changes in v4.

javierguerragiraldez commented on README.md:
```diff
@@ -12,16 +12,16 @@ The current stable version is [*v4*](v4/) and is fully go-mod compatible.
+$ go get github.com/mikefero/semver/v4
 # Or use fixed versions
-$ go get github.com/blang/semver/v4@v4.0.0
+$ go get github.com/mikefero/semver/v4@v4.0.0
```
i guess it's now
```suggestion
$ go get github.com/Kong/semver/v4@v4.0.0
```

javierguerragiraldez commented on package.json:
```diff
@@ -1,17 +1,20 @@
-    "URL": "https://github.com/blang/semver/issues",
-    "url": "https://github.com/blang/semver/issues"
+    "URL": "https://github.com/mikefero/semver/issues",
+    "url": "https://github.com/mikefero/semver/issues"
```
again `s/mikefero/Kong/g`

mikefero replied:
That's what I get for starting on my fork ;).

javierguerragiraldez commented on range.go:
```diff
@@ -301,6 +304,19 @@ func incrementMinorVersion(vStr string) (string, error) {
+	parts := strings.Split(vStr, ".")
+	i, err := strconv.Atoi(parts[2])
+	if err != nil {
+		return "", err
```
this error could be wrapped to give some context

mikefero replied:
Wrapped this and all the other increment functions in range.go

javierguerragiraldez commented on range_test.go:
```diff
@@ -295,61 +345,125 @@ func TestExpandWildcardVersion(t *testing.T) {
-	if !rf(MustParse("1.2.2")) || rf(MustParse("1.2.3")) {
-		t.Errorf("Invalid conversion to range func")
-	}
+	t.Run("3 digit ranges", func(t *testing.T) {
```
just curious, does the use of `t.Run` raises the minimum Go version usable?

i know the Go community is quite fast to update the compiler, so it shouldn't be a problem, unless it's declared in some machine-readable file.

mikefero replied:
I am not sure, I will have to look into that one, the version listed in go.mod is 1.12 if I remember correctly.

mikefero replied:
Looks like it was added in [go 1.7](https://pkg.go.dev/testing#T.Run) so we should be good here; good thinking.

javierguerragiraldez commented on semver.go:
```diff
@@ -11,32 +12,78 @@ const (
+	// version numbers. There are six capture groups which define major, minor, patch, revision,
+	// pre-release, and build metadata. When using 3 digit versioning strict semver adherence
+	// will be utilized/
+	threeFourDigitVersionRegexFormat = `^(?P<major>0|[1-9]\d*)\.` +
```
i'm never sure about operator precedence in regexes.  is `0|[1-9]\d*` equivalent to `(0)|([1-9]\d*)` or `(0|[1-9])\d*`? if the latter, it would allow numbers like `01`.

mikefero replied:
So I modified this from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string adding the revision and created the tolerant variant too.

javierguerragiraldez replied:
ok, so i tried it: https://go.dev/play/p/ANozC6L9sJM with surprising results:
- `^0|[1-9]\d*$` accepts "01".  not good
- `^0|([1-9]\d*)$` also accepts "01".... why?
- `^(0|[1-9]\d*)$`  does reject "01"  so it seems these are fine.

found the explanation in [a POSIX doc](https://www.boost.org/doc/libs/1_56_0/libs/regex/doc/html/boost_regex/syntax/basic_extended.html#boost_regex.syntax.basic_extended.operator_precedence).   in short: alternation '|' is the lowest precedence of all, which is good, but it's lower than anchors (^ and $).

the second pattern essentially means "starts with 0 or ends with digits", which isn't useful.  the last pattern is basically: "full string: 0 or nice number"

TL;DR: these patterns are ok because each alternation is wrapped in some group

javierguerragiraldez commented on semver.go:
```diff
@@ -11,32 +12,78 @@ const (
 }
 
-// Version to string
+// KongVersion to string
```
what is a `KongVersion`?  (and how is it kongverted ? :laughing:  )

mikefero replied:
It is all the rage and it converts to to all things; that's just how it rolls.

javierguerragiraldez commented on semver.go:
```diff
@@ -64,12 +111,16 @@ func (v Version) String() string {
 // FinalizeVersion discards prerelease and build number and only returns
 // major, minor and patch number.
```
... and optionally revision

javierguerragiraldez commented on semver.go:
```diff
@@ -173,16 +234,30 @@ func (v Version) Compare(o Version) int {
+	if v.Revision >= 0 {
+		v.Revision++
+	}
+	return nil
```
this will silently be ignored if there wasn't a revision number.  i'd guess calling this should count as "first revision".  otherwise, return an error.

or am i misinterpreting the semantics?   if revision>=0 means "i do use revisions" while revision==-1 means "this is a real semver", then this function should fail on revision==-1.

mikefero replied:
So `-1` means revision is disabled and there should be no changes to revision if you call it; if there is a revision >=0 then increment. If revision is not parsed then we should make sure it never gets incremented or else it could get added to the string.

mikefero replied:
I updated the godoc for this function and made it a little more descriptive; let me know if this clears things up.

javierguerragiraldez commented on semver.go:
```diff
@@ -238,31 +316,74 @@ func Make(s string) (Version, error) {
-			}
-			parts[i] = p
+	if !TolerantVersionRegex.MatchString(s) {
+		return Version{}, fmt.Errorf("Invalid tolerant version: '%s'", s)
```
nit: the "tolerant" word sounds weird here, as it's not the version the tolerant here, it's the parser.  and i don't think it adds any meaning.  not for the developer (who knows which parser is being used) nor for the user (who shouldn't care).

(subnit: thou shall not capitalise errors)

mikefero replied:
Yeah I don't usually capitalize errors, but this module is riddled with bad practices so I kept things this way to be consistent.

mikefero replied:
Updating the name of the regex to use the word "permissive" and also updating the poorly written log message by removing `tolerant`.

javierguerragiraldez commented on semver.go:
```diff
@@ -238,31 +316,74 @@ func Make(s string) (Version, error) {
+	}
+	if len(buildStr) > 0 {
+		s = fmt.Sprintf("%s+%s", s, buildStr)
+	}
```
what about `strings.Builder`?

GGabriele commented on semver.go:
```diff
@@ -11,32 +12,78 @@ const (
+		`(?P<patch>0|[1-9]\d*)` +
+		`(?:\.(?P<revision>0|[1-9]\d*))?` +
+		`(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?` +
+		`(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`
```
Mother of regex :D

GGabriele commented on semver.go:
```diff
@@ -11,32 +12,78 @@ const (
+	// pre-release, and build metadata. The loose format allows for leading and trailing spaces,
+	// prefixing versions with a 'v' and empty capture groups in order to dynamically fill in the
+	// fields with defaults. When using 3 digit versioning loose semver adherence will be utilized.
+	threeFourDigitTolerantVersionRegexFormat = `^(?:\s*)?(?:v)?` +
```
Maybe using `Loose` instead of `Tolerant`? The latter doesn't sound right, but maybe it's because of my not-native-english ears :P

mikefero replied:
Indeed. I chose tolerant because of the `ParseTolerant()` method call, but I think that makes more sense as the function name and action that it performs; loose or some form of that does sound more appealing.

mikefero replied:
Updating the name to permissive and also updating the documentation to give more context.

javierguerragiraldez commented on semver.go:
```diff
@@ -374,18 +379,26 @@ func ParseTolerant(s string) (Version, error) {
 
 	// Generate version to properly parse
-	s = fmt.Sprintf("%s.%s.%s", majorStr, minorStr, patchStr)
+	b := strings.Builder{}
+	b.WriteString(majorStr)
+	b.WriteString(".")
+	b.WriteString(minorStr)
+	b.WriteString(".")
+	b.WriteString(patchStr)
```
```suggestion
	fmt.Fprintf(&b, "%s.%s.%s", majorStr, minorStr, patchStr)
```

javierguerragiraldez commented on semver.go:
```diff
@@ -374,18 +379,26 @@ func ParseTolerant(s string) (Version, error) {
+	b.WriteString(patchStr)
 	if len(revisionStr) > 0 {
-		s = fmt.Sprintf("%s.%s", s, revisionStr)
+		b.WriteString(".")
+		b.WriteString(revisionStr)
```
```suggestion
		fmt.Fprintf(&b, "%s.%s", s, revisionStr)
```

mikefero replied:
This is adding the full version string back in to the string builder; pretty sure we'll need to keep the `WriteString` calls here to build the rest of the string.

javierguerragiraldez replied:
oh, right, it should be `fmt.Fprintf(&b, ".%s", revisionStr)`

but of course, to replace just two `WriteString`s it's not that much of an improvement

mikefero commented:
I am going to remove Travis here as well and make a very simplistic workflow for GHA.

mikefero committed:
feat: add Makfile and GitHub actions

This adds a simple Makefile for testing and building/running the
examples for semver. This also removes TravisCI and replaces it with
a GitHub action in order to execute the unit tests.
